p2panda-net/src/engine/topic_discovery.rs

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use anyhow::Result;
+use iroh_net::NodeId;
+use rand::random;
+use serde::{Deserialize, Serialize};
+use tokio::sync::mpsc;
+
+use crate::bytes::{FromBytes, ToBytes};
+use crate::engine::address_book::AddressBook;
+use crate::engine::constants::JOIN_PEERS_SAMPLE_LEN;
+use crate::engine::gossip::ToGossipActor;
+use crate::NetworkId;
+
+#[derive(Default, PartialEq, Eq)]
+enum Status {
+    #[default]
+    Idle,
+    Pending,
+    Active,
+}
+
+/// Manages the "topic discovery" background process.
+///
+/// Within a network peers can be interested in different topics. How topics are defined is up to
+/// the applications. Within a network there can be even multiple applications co-existing.
+///
+/// To find out which peer is interested in what topic we need a process called "topic discovery".
+/// Currently this is (rather naively) implemented as a network-wide gossip overlay where peers
+/// frequently broadcast their interests. Later we might look into other approaches, for example
+/// applying a random-walk algorithm which traverses the network and learning about it over time.
+// @TODO(adz): Would be great to already express this interface as traits so it's easier to swap
+// out the strategies with something else. The API could even look similar to our current
+// `Discovery` trait (for peer discovery), adjusted to work with topics.
+pub struct TopicDiscovery {
+    address_book: AddressBook,
+    gossip_actor_tx: mpsc::Sender<ToGossipActor>,
+    network_id: NetworkId,
```

Comment:
I wonder if we could get rid of the `network_id` field here, since `address_book` already includes the network id internally? You could add a `network_id()` method to `AddressBook` for convenient look-up.

Reply:
I think intuitively I wouldn't understand why the address book is the source for knowing about the network id (it _also_ needs to know it but I wouldn't see it as the solely source)

Reply:
Agreed. That makes sense to me!